### ISSUE TITLE
keras_saved_model fails becase model is not json serializable

### ISSUE THREAD

**Issue by agdownes** [tensorflow/probability#325]
tf 1.13
tfp 0.6
ubuntu 18.04

Trying to save a Bayesian neural net in a way that can be run in production using either the java or c++ api. I've tried 

`save_keras_model(model,
                 saved_model_path='model_path', serving_only=True,
                 custom_objects={'DenseFlipout': tfp.layers.DenseFlipout,
                                 'Conv2DFlipout': tfp.layers.Convolution2DFlipout})`

and just:
`save_keras_model(model,
                 saved_model_path='model_path', serving_only=True)`

Both fail because model.to_json() fails

This minimal example produces the below error:

`import tensorflow as tf
import tensorflow_probability as tfp
from tensorflow.contrib.saved_model import save_keras_model

tf.enable_eager_execution()

neural_net = tf.keras.Sequential([
        tfp.layers.Convolution2DFlipout(6,
                                        input_shape=(3, 255, 255),
                                        kernel_size=5,
                                        padding="SAME",
                                        activation=tf.nn.relu),
        tf.keras.layers.MaxPooling2D(pool_size=[2, 2],
                                     strides=[2, 2],
                                     padding="SAME"),
        tfp.layers.Convolution2DFlipout(16,
                                        kernel_size=5,
                                        padding="SAME",
                                        activation=tf.nn.relu),
        tf.keras.layers.MaxPooling2D(pool_size=[2, 2],
                                     strides=[2, 2],
                                     padding="SAME"),
        tfp.layers.Convolution2DFlipout(120,
                                        kernel_size=5,
                                        padding="SAME",
                                        activation=tf.nn.relu),
        tf.keras.layers.Flatten(),
        tfp.layers.DenseFlipout(84, activation=tf.nn.relu),
        tfp.layers.DenseFlipout(10)
])

save_keras_model(neural_net, saved_model_path='model_path', serving_only=True)
`

  File "/home/tony/.conda/envs/tf/lib/python3.6/site-packages/tensorflow/contrib/saved_model/python/saved_model/keras_saved_model.py", line 141, in save_keras_model
    _export_model_json(model, export_dir)
  File "/home/tony/.conda/envs/tf/lib/python3.6/site-packages/tensorflow/contrib/saved_model/python/saved_model/keras_saved_model.py", line 151, in _export_model_json
    model_json = model.to_json()
  File "/home/tony/.conda/envs/tf/lib/python3.6/site-packages/tensorflow/python/keras/engine/network.py", line 1563, in to_json
    return json.dumps(model_config, default=get_json_type, **kwargs)
  File "/home/tony/.conda/envs/tf/lib/python3.6/json/__init__.py", line 238, in dumps
    **kw).encode(obj)
  File "/home/tony/.conda/envs/tf/lib/python3.6/json/encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/home/tony/.conda/envs/tf/lib/python3.6/json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "/home/tony/.conda/envs/tf/lib/python3.6/site-packages/tensorflow/python/keras/engine/network.py", line 1560, in get_json_type
    raise TypeError('Not JSON Serializable:', obj)
TypeError: ('Not JSON Serializable:', <function default_loc_scale_fn.<locals>._fn at 0x7f8a640f2378>)

Are there any tested ways of taking tensor_flow probability to production other than running pure python?

**Comment by agdownes** [comment]
I have also tried exporting just the protobuf ModelDef using https://stackoverflow.com/questions/45466020/how-to-export-keras-h5-to-tensorflow-pb

This works for me using a regular tensorflow model, I can load and run the saved model in Java. It also seems to work when saving the Bayesian neural net but I get a null pointer exception when trying to run the model in java.

**Comment by jvdillon** [comment]
Hi @agdownes. Sorry you ran into this issue--this is a feature we know is missing. We'll try to add it as soon as we can!

**Comment by Sourav-Bhattacharya10** [comment]
Hi.
I have created this model:
```shell
import tensorflow as tf
import tensorflow_probability as tfp
tfd = tfp.distributions

model = tf.keras.Sequential([
  tf.keras.layers.Dense(1),
  tfp.layers.DistributionLambda(lambda t: tfd.Normal(loc=t, scale=1)),
])

curr_optimizer = tf.keras.optimizers.Adam()
    
negloglik = lambda y, rv_y: -rv_y.log_prob(y)
    
model.compile(optimizer=curr_optimizer, loss=negloglik, metrics=['accuracy']
```
I saved it using:
```shell
tf.keras.models.save_model(model, 'Model4.h5')
```
When I try to load it using:
```shell
model = tf.keras.models.load_model('Model4.h5')
```

It gives this error:

```shell
ValueError: Unknown layer: DistributionLambda
```
Is it not possible to load tf.Keras models created using tensorflow_probability?

**Comment by janosh** [comment]
Having the same problem. Looking forward to this functionality being added. 👍

**Comment by bglick13** [comment]
For what it's worth, this seems to be a decent workaround...

to save:
`
file = h5py.File('models/{}.h5'.format(model_name), 'w')
weight = model.get_weights()
for i in range(len(weight)):
   file.create_dataset('weight' + str(i), data=weight[i])
file.close()
`

to load:
`
file = h5py.File('models/{}.h5'.format(model_name), 'r')
weight = []
for i in range(len(file.keys())):
   weight.append(file['weight' + str(i)][:])
model.set_weights(weight)
`

**Comment by agdownes** [comment]
@bglick13 Yup that should work. I also found using model checkpoints with weights_only=true works too, then just loading the weights using model.load_weights(weights_file). So presumably just plain model.save_weights also works, I have not tested this but it would reduce the saving to a single line of code. This works for me as I don't usually care about saving the optimizer state. 
  
The reason I opened this issue was for the use of "save_keras_model" which saves into the Tensorflow protobuf format and can then be used in TensorFlow Extended or Serving for use in production. For this cases just saving the model weights in h5 is not a work around unfortunately.

The case of just saving the keras model in h5 format with model.save was raised in issue #289  https://github.com/tensorflow/probability/issues/289

**Comment by jankislinger** [comment]
If you are having problem with loading the model that produces this error
```
ValueError: Unknown layer: DistributionLambda
```

You could find a workaround by importing the layer explicitly.
```
from tf.keras import models
from tensorflow_probability.layers import DistributionLambda

model = models.load_model('Model4.h5')
```

**Comment by Himscipy** [comment]
@jvdillon Any updates on saving the BNN model for reuse/modification? @jankislinger I tried your approach but it didn't worked for me. @Sourav-Bhattacharya10 Did you found any work around till now. ? Thank you

**Comment by nbro** [comment]
@jvdillon Any updates? Currently, I am getting the following warning when training a model with TFP's probabilistic layers and using TensorBoard to visualize the evolution of the loss

> WARNING:tensorflow:Model failed to serialize as JSON. Ignoring... ('Not JSON Serializable:', <tf.Variable 'conv2d_flipout/kl_loss_weight_copy:0' shape=() dtype=float32, numpy=1.0>)

This feature is really essential. Is it possible to at least know when do you plan to add this feature?

**Comment by nbro** [comment]
> Hi.
> I have created this model:
> 
> ```shell
> import tensorflow as tf
> import tensorflow_probability as tfp
> tfd = tfp.distributions
> 
> model = tf.keras.Sequential([
>   tf.keras.layers.Dense(1),
>   tfp.layers.DistributionLambda(lambda t: tfd.Normal(loc=t, scale=1)),
> ])
> 
> curr_optimizer = tf.keras.optimizers.Adam()
>     
> negloglik = lambda y, rv_y: -rv_y.log_prob(y)
>     
> model.compile(optimizer=curr_optimizer, loss=negloglik, metrics=['accuracy'])
> ```
> 
> I saved it using:
> 
> ```shell
> tf.keras.models.save_model(model, 'Model4.h5')
> ```
> 
> When I try to load it using:
> 
> ```shell
> model = tf.keras.models.load_model('Model4.h5')
> ```
> 
> It gives this error:
> 
> ```shell
> ValueError: Unknown layer: DistributionLambda
> ```
> 
> Is it not possible to load tf.Keras models created using tensorflow_probability?

If I execute this code with TF 2.1 and TFP 0.9, I actually get another error.

> ValueError: Weights for model sequential have not yet been created. Weights are created when the Model is first called on inputs or `build()` is called with an `input_shape`.

which can be solved by specifying the `input_shape` parameter of the only `Dense` layer, that is, for example, something like this.

```
tf.keras.layers.Dense(1,input_shape=(1,))
```
After that, I don't get the error mentioned in this comment. Maybe because I am using a different version of TF and TFP that it was used by @Sourav-Bhattacharya10.

**Comment by waldnerf** [comment]
This [post](https://stackoverflow.com/a/60621157/1948347) worked for me
